src/basedtyping/__init__.py

```diff
@@ -591,6 +587,116 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
+            try:
+                ast.parse(arg_to_compile, mode="func_type")
+            except SyntaxError:
+                raise SyntaxError(f"invalid syntax in ForwardRef: {arg_to_compile}?") from None
+            else:
+                code = compile("'un-representable callable type'", "<string>", "eval")
+
+        self.__forward_arg__ = arg
+        self.__forward_code__ = code
+        self.__forward_evaluated__ = False
+        self.__forward_value__ = None
+        self.__forward_is_argument__ = is_argument
+        self.__forward_is_class__ = is_class
+        self.__forward_module__ = module
+
+    if sys.version_info >= (3, 13):
+
+        @override
+        def _evaluate(
+            self,
+            globalns: dict[str, object] | None,
+            localns: dict[str, object] | None,
+            type_params: tuple[TypeVar | typing.ParamSpec | typing.TypeVarTuple, ...] = (),
+            *,
+            recursive_guard: frozenset[str],
+        ) -> object | None:
+            return transformer._eval_direct(self, globalns, localns)
+
+    elif sys.version_info >= (3, 12):
+
+        @override
+        def _evaluate(
+            self,
+            globalns: dict[str, object] | None,
+            localns: dict[str, object] | None,
+            type_params: tuple[TypeVar | typing.ParamSpec | typing.TypeVarTuple, ...] | None = None,
+            *,
+            recursive_guard: frozenset[str],
+        ) -> object | None:
+            return transformer._eval_direct(self, globalns, localns)
+
+    else:
+
+        @override
+        def _evaluate(
+            self,
+            globalns: dict[str, object] | None,
+            localns: dict[str, object] | None,
+            recursive_guard: frozenset[str],
+        ) -> object | None:
+            return transformer._eval_direct(self, globalns, localns)
+
+
+def _type_check(arg: object, msg: str) -> object:
+    """Check that the argument is a type, and return it (internal helper).
+
+    As a special case, accept None and return type(None) instead. Also wrap strings
+    into ForwardRef instances. Consider several corner cases, for example plain
+    special forms like Union are not valid, while Union[int, str] is OK, etc.
+    The msg argument is a human-readable error message, e.g::
+
+        "Union[arg, ...]: arg should be a type."
+
+    We append the repr() of the actual value (truncated to 100 chars).
+    """
+    invalid_generic_forms = (Generic, typing.Protocol)
+
+    arg = _type_convert(arg)
+    if isinstance(arg, _GenericAlias) and arg.__origin__ in invalid_generic_forms:  # type: ignore[comparison-overlap]
+        raise TypeError(f"{arg} is not valid as type argument")
+    if arg in (Any, NoReturn, typing.Final, Untyped):
+        return arg
+    if isinstance(arg, _SpecialForm) or arg in (Generic, typing.Protocol):
+        raise TypeError(f"Plain {arg} is not valid as type argument")
+    if isinstance(arg, (type, TypeVar, ForwardRef)):
+        return arg
```

Comment:
Will this work with `typing_extensions`?

Reply:
i'm gonna assume yes, those defs are the ones from `typing_extensions`

Reply:
Ah yes, `typing_extensions.TypeVar` doesn't subclass `typing.TypeVar`, but does some sketchy metaclass `__instancecheck__` black voodoo magic (whilst forgetting about `__subclasscheck__` BTW). 
Because why the hell would the elite `typing_extensions` play by their own typing rules, right?